### ISSUE TITLE
[Bug] Abnormally large `grad_norm` in PaTH training when `cu_seqlens != None`

### ISSUE THREAD

**Issue by lxr-tech** [fla-org/flash-linear-attention#837]
### Checklist

- [x] I have checked [FAQs](https://github.com/fla-org/flash-linear-attention/blob/main/FAQs.md) and existing issues for similar problems
- [x] Please report this bug in English to ensure wider understanding and support

### Describe the Bug

In our PaTH pretraining runs, `grad_norm` becomes abnormally large when variable-length mode is enabled (`cu_seqlens != None`).

At the same time:
- loss can still decrease,
- and if we switch to fixed-length mode (`cu_seqlens = None`), `grad_norm` immediately looks normal.

This issue is reported as a reproducible training instability centered on `grad_norm`.

Could you please help check whether this behavior is expected in PaTH varlen mode, and whether an upstream fix is needed?

### Steps to Reproduce the Bug

Under the same codebase and training setup:
1. `cu_seqlens = None` (fixed-length path): training is stable and `grad_norm` stays in a normal range.
2. `cu_seqlens != None` (varlen path): `grad_norm` is significantly larger.
3. If we manually pack `[B, T, ...] -> [1, B*T, ...]` before PaTH varlen attention and unpack outputs afterward, both loss and `grad_norm` become normal again.

For quick location: in `fla/layers/path_attn.py`, the training path gets `batch_size, q_len` from `[B, T, ...]` and directly calls `parallel_path_attn(..., cu_seqlens=cu_seqlens)` (see https://github.com/fla-org/flash-linear-attention/blob/main/fla/layers/path_attn.py#L118).

### Expected Behavior

See description.

### Environment Information

- flash-linear-attention commit: `f24317a6a4f513748cd7eb05818534ce66029957`
- PyTorch: `2.6.0`
- Triton: `3.2.0`
- CUDA: `12.4`
- GPU: `NVIDIA H200`

**Comment by zhiyuan1i** [comment]
Thanks, I’ve recreated it and will take a look.

**Comment by zhiyuan1i** [comment]
Hi, thanks for the detailed report.

This is not a kernel bug, but rather a usage issue: **when using `cu_seqlens`, the batch size `B` must be 1**. The variable-length sequences should be flattened/packed into shape `[1, total_len, ...]` before passing to the op, consistent with the FlashAttention API.

In your reproduction, `q`/`k`/`v` still have shape `[B, T, ...]` with `B > 1` while `cu_seqlens` is provided. This causes the Triton kernels to schedule duplicate grids across the batch dimension, leading to the abnormal `grad_norm` you observed. When you manually pack to `[1, B*T, ...]` (your "packed" test), everything works correctly because the batch-size requirement is satisfied.

We have now added explicit `ValueError` checks at the public API entry points (including `parallel_path_attn`) so that users get a clear error message instead of silently incorrect gradients when `B != 1` with `cu_seqlens`. See the linked PR for the fix.

**Correct usage:**
```python
# Pack [B, T, ...] -> [1, B*T, ...]
x = x.view(1, B * T, H * D)
cu_seqlens = torch.cumsum(torch.tensor([0] + [T] * B, device=x.device, dtype=torch.int32), dim=0)
out, _, _ = model(x, cu_seqlens=cu_seqlens)
```

Closing this as the root cause has been identified and fixed. Feel free to reopen if you still see issues after packing the inputs correctly.

**Comment by lxr-tech** [comment]
Hi, thanks for the clarification and for the update in PR #844.

I understand the `B == 1` requirement with `cu_seqlens`, and our local packing fix works well on my side.

One small note: in HuggingFace models, Q/K/V are typically in `[B, T, H, D]` shape, so keeping good compatibility with this common layout is still practically useful for users.

For now, I will continue with our local adjusted version. I will also try the updated upstream code later.

Thanks again for the quick support.
